ISSUE-22: Documentation for XDebug, PHPStorm, and Archipelago's Docker

marlo-longley opened the pull request:
This pull addresses #22 . It adds images and a markdown file explaining the process of installing Xdebug on PHPStorm using it with Archipelago's php docker containers.

To build the esmero-php-fpm image with Xdebug installed, see [this PR ](https://github.com/esmero/archipelago-docker-images/pull/4) in the `archipelago-docker-images` repository.

@DiegoPino 
Ok. This could be less complicated if we had our own :development tagged php image in docker hub. Then we could hide the whole first step of this document (Part 1: Docker) from the user. Except, we need them to input their local machine IP. So that's how I left it for now. Will link to corresponding pull in archipelago-docker-images so you can build.

marlo-longley committed:
Add documentation for using XDebug, PHPStorm, and Archipelago's Docker setup

DiegoPino commented:
@marlo-longley we do:
https://github.com/esmero/archipelago-docker-images/tree/master/esmero-php-fpm
We just need to update that image and publish with a new tag
XDEBUG can be installed, but for production disabled on the .ini file

EDIT: you mean we have that not published yet. Well of course. That is why you are doing this! But we will once i test the steps.

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,68 @@
+
+## Part 2: PHPStorm
+
+1. In PHPStorm, open your `archipelago-deploymenet` project.
```
small typo `archipelago-deploymenet` = `archipelago-deployment`

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,68 @@
+1. Find out what your local machine (your actual computer!) IP address is. For example mine is, `192.168.1.3`
+
+1. Build the development image with Xdebug enabled. Run the following from the `/archipelago-docker-images/esmero-php-fmp` folder.
+`sudo docker build -f Dockerfile.dev --build-arg LOCAL_IP=<your machine IP> -t esmero-php:development . --no-cache`
```
Can we pass the machine IP as a config option instead of during build?

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,68 @@
+1. In PHPStorm, open your `archipelago-deploymenet` project.
+ 
+2. Go to `Preferences > Languages & Frameworks > PHP > Debug`. In this window there is an Xdebug section. Use these settings:
+    - Debug port: `9001`. (do NOT use the default, 9000)
```
Not 9000 Because we have min.io right? maybe we could push it to 9999? Just to be sure we don't end using the next number for a service in the future?

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,68 @@
+      ![Debug](../imgs/xdebug/telephone.png) 
+ 2. Now, you can use `Run > Debug` and select the `docker-debug-server` that we created in the previous steps. The debugging console will appear. Happy debugging!
+ 
+ ####Questions? You can ask Marlo at  [mlongley@metro.org](mlongley@metro.org) or post on our [Google Group](https://groups.google.com/forum/#!forum/archipelago-commons)
```
jajaja.Nice. IMHO, Let's try to avoid our personal emails, people will always prefer that (of course!!) and you will end answering things over and over and over!

DiegoPino reviewed (commented):
Good work 🎉 . Will give you my full review by tomorrow =) In the meantime some tiny small things

DiegoPino commented:
@marlo-longley when you make a pull request please explain what the pull does and also link/mention via the automatic hashing system that github provides which ISSUE this resolves.

DiegoPino commented:
@mitchellkeaney so i don't have to repeat over and over how to make a pull, do you think you could take my contributing guideline and make it a github template? https://help.github.com/en/github/building-a-strong-community/creating-a-pull-request-template-for-your-repository

mitchellkeaney commented:
hey @DiegoPino - thank you for bringing this to my attention, but at the moment I do not have the bandwidth to complete your request. can i send this back to you and in the coming weeks work with you to create a small list of integral documentation that needs to be created/updated?

thank you!!

DiegoPino commented:
@mitchellkeaney all good. I will deal with this. Thanks

DiegoPino commented:
@marlo-longley please review my comments on https://github.com/esmero/archipelago-docker-images/pull/4. I feel that will dictate the outcome of this pull. Thanks

marlo-longley committed:
Address review comments

marlo-longley committed:
Update Xdebug workflow instructions

marlo-longley commented:
@DiegoPino I addressed a lot of your concerns. 
See the updated Dockerfile PR [here](https://github.com/esmero/archipelago-docker-images/pull/4).

Also, see the [PR](https://github.com/esmero/archipelago-deployment/pull/21) to add `docker-compose.dev.yml` file (referenced in this documentation which uses an override with that file to launch the dev/debug image).

The main thing still is the logging. I added a workaround method to get the xdebug.log into the esmero-php docker logs in step 4 of the final section of this document. It is not the most great solution but I wanted to get something published and move forward.

I was also able to avoid problems with nginx after using this setup for a couple days. However, if I leave a multiple breakpoints hanging open on a request for a long time, I get 504 Gateway Timeout eventually. I think people could adjust this to their work needs and is OK for now.

I am curious for you to try it out. I hope it will help us be faster. Thank you!

marlo-longley commented:
@mitchellkeaney @DiegoPino About the github templates for Issues and PRs - I am going to open an issue for this and address it. I was working on it already a bit today because it came back to my attention from this thread. More soon.

DiegoPino commented:
@marlo-longley you updating this documentation to match https://github.com/esmero/archipelago-docker-images/pull/4? Lets work on that 2020

marlo-longley commented:
@DiegoPino Wow yes, I had updated it to go along with my work, but didn't push! Here it is updated now.

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,85 @@
+This document describes how to enable Xdebug for local PHP development using the PHPStorm IDE and a docker container running the Archipelago `esmero-php:development` image. It involves interacting with the [esmero/archipelago-docker-images](https://github.com/esmero/archipelago-docker-images) repo and the [esmero/archipelago-deployment](https://github.com/esmero/archipelago-deployment) repo.
+
+## Part 1: Docker
+1. Build the development image with Xdebug enabled. Run the following from the `/archipelago-docker-images/esmero-php-fmp` folder:
```
@marlo-longley wondering if we could leave the building part as an extra/advanced item and simply deliver built container (build by us) from docker? What do you think? I feel it is super good to have the "how" a container is build, i really like it, but also, i would love to lower the barrier a bit. Lets talk about this on Monday! Happy new year and thanks for all this good work =)

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,85 @@
+1. In PHPStorm, open your `archipelago-deployment` project.
+ 
+2. Go to `Preferences > Languages & Frameworks > PHP > Debug`. In this window there is an Xdebug section. Use these settings:
+    - Debug port: `9999`. (do NOT use the default, 9000)
```
Good!

DiegoPino commented on docs/xdebug.md:
```diff
@@ -0,0 +1,85 @@
+ 3. Go to your Archipelago at `localhost:8001` and refresh the page. If you have breakpoints set, either manually, or from leaving "Break at first line in PHP scripts" checked, you should have output now in the debugger. If you have no breakpoints set, at least the "Waiting for incoming" message will now say "Connected".
+ 
+ 4. If you would like to see the output of your xdebug logs, run the following script:
+ `docker exec -ti esmero-php bash -c 'tail -f /tmp/xdebug.log > /proc/1/fd/2'`
```
I feel this is correct. I know i said make the logs part of the container output, but then looking at this it feels better. That way we can separate logs coming from xdebug.log from the general phpfpm ones. But, a catch here. Do we need a cron that clears those logs? Just a question since i have not tested this, just worried the container local storage will end being full of logs after a few days of operation. Any ideas on that?

DiegoPino reviewed (approved):
Looking good!

DiegoPino reviewed (approved):
Good. Merging. We can refine as we go deeper into this. I will build and publish the esmero-php image tagged for development later today!
